### PR TITLE
Use React 17 types in DCR

### PR DESCRIPTION
**Pull request by joecowton1**
## What does this change?
Rolls back `@types/react` to v17 and rebuilds the `package.lock`. 

## Why?
We've encountered all sorts of problems lately trying to bump packages, mostly caused by DCR using v18 of `@types/react`. By reseting the `package.lock` and rolling back to `"@types/react": "^17.0.52",` we should be resolve those issues.

### Other stuff caught by CI
- increases GA bundle package size limit
- updates out of date browser targets
- runs prettier on files that needed formatting
- adds missing keys to mapped components

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,6 +26,7 @@
     "prepare": "husky install"
   },
   "resolutions": {
+    "@types/react": "^17.0.52",
```
Comment (SiAdcock) [inline package.json:29]: Is this still necessary?

**Comment by SiAdcock** [comment]
It's encouraging to see the build passing. The consequences of the changes to `yarn.lock` are difficult to imagine. It's great that we're shaving 446KB off the bundle size! Is it the right 446KB? 😅

File: dotcom-rendering/cypress/support/commands.js [hunk 2]
```diff
@@ -28,19 +28,29 @@ import 'cypress-wait-until';
...
 		.each((el) => {
-			cy.wrap(el)
-				.log(`Scrolling to ${el.attr('name')}`)
-				.scrollIntoView({ duration: 100, timeout: 10000 })
-				.should('have.attr', 'data-gu-ready', 'true', {
-					timeout: 30000,
-				});
+			const deferuntil = el.attr('deferuntil');
+			const name = el.attr('name');
+			const defer = el.attr('deferuntil');
```
Comment (mxdvl) [inline dotcom-rendering/cypress/support/commands.js:35]: Bit of repetition there!

File: dotcom-rendering/cypress/support/commands.js [hunk 3]
```diff
@@ -28,19 +28,29 @@ import 'cypress-wait-until';
...
+			const islandMeta = `island: ${name} defer: ${defer}`;
+			if (['idle', 'visible', undefined].includes(deferuntil)) {
+				cy.log(`Scrolling to ${islandMeta}`);
+				cy.wrap(el)
+					.scrollIntoView({ duration: 1000, timeout: 30000 })
+					.should('have.attr', 'data-gu-ready', 'true', {
+						timeout: 30000,
```
Comment (mxdvl) [inline dotcom-rendering/cypress/support/commands.js:42]: Small readability improvement:

```suggestion
					.scrollIntoView({ duration: 1000, timeout: 30_000 })
					.should('have.attr', 'data-gu-ready', 'true', {
						timeout: 30_000,
```

File: dotcom-rendering/scripts/webpack/browser-targets.test.ts [hunk 4]
```diff
@@ -11,7 +11,7 @@ describe('Browser targets are as expected', () => {
 			edge: '91.0.0',
 			firefox: '78.0.0',
 			ios: '10.3.0',
-			opera: '64.0.0',
+			opera: '88.0.0',
```
Comment (mxdvl) [inline dotcom-rendering/scripts/webpack/browser-targets.test.ts:14]: Is this meant to be part of this PR?

File: package.json (hunk 1 above)
Comment (mxdvl) [reply package.json:29]: Is this because there could be other versions coming as transitive dependencies?

**Review by mxdvl — COMMENTED** [review]
Small comments below, but this is great and we should enforce a stricter `@types/*` package policy to prevent this in the future!

File: dotcom-rendering/cypress/support/commands.js (hunk 3 above)
Comment (arelra) [reply dotcom-rendering/cypress/support/commands.js:41]: This causes the lint error `Parsing error: Identifier directly after number` so I have removed for now.

File: package.json (hunk 1 above)
Comment (joecowton1) [reply package.json:29]: Yes, without this we get a mismatch with `react-google-recaptcha` which uses types v18

File: dotcom-rendering/scripts/webpack/browser-targets.test.ts (hunk 4 above)
Comment (joecowton1) [reply dotcom-rendering/scripts/webpack/browser-targets.test.ts:14]: Yeah, for some reason reseting the package.lock means these are caught in failing tests.

**Comment by joecowton1** [comment]
> It's encouraging to see the build passing. The consequences of the changes to `yarn.lock` are difficult to imagine.
> 
> For instance, it's great that we're shaving 446KB off the bundle size! Is it the right 446KB? 😅

That seems like a huge amount right. I think the situation we're in currently - were for example bumping libs became almost impossible without doing something like this - has forced our hand slightly though: either we continue with a package.lock thats a mess and doesn't allow us to add packages, or we do a hard reset and make a real effort to stay on top of things from here on.

**Comment by SiAdcock** [comment]
> either we continue with a package.lock thats a mess and doesn't allow us to add packages, or we do a hard reset and make a real effort to stay on top of things from here on.

I feel you, it's likely we have reached that point. I'm happy if WebX are, of course, just wanted to flag it so we can have the conversation 😊 

Some of these dist files are being completely removed, the ones with the opaque-sounding names e.g.

```
dotcom-rendering/dist/9808.legacy.********************.js
```

I wonder if they were just not being used, were hanging around for some gnarly package incompatibility reasons that the yarn.lock reset has shaken out? Anyway, I'm trying to say, maybe the huge drop in bytesize is an anomaly.

**Comment by mxdvl** [comment]
> Some of these dist files are being completely removed, the ones with the opaque-sounding names e.g.
> 
> ```
> dotcom-rendering/dist/9808.legacy.********************.js
> ```
> 
> I wonder if they were just not being used, were hanging around for some gnarly package incompatibility reasons that the yarn.lock reset has shaken out? Anyway, I'm trying to say, maybe the huge drop in bytesize is an anomaly.

Yeah you cannot trust the [bundle size action](https://github.com/guardian/dotcom-rendering/pull/6967#issuecomment-1384248171) for unnamed/numeric chuncks. This happens every now and again when Wepback chooses a different chunking strategy, and the bot only reports removal, but not addition, so you cannot really rely on its report. Here, we are actually adding 30 kB (0.03 MB) if you look at the total size of recent PRs (2.94 MB v 2.97 MB)

**Comment by joecowton1** [comment]
> Some of these dist files are being completely removed, the ones with the opaque-sounding names e.g.
> 
> ```
> dotcom-rendering/dist/9808.legacy.********************.js
> ```
> 
> I wonder if they were just not being used, were hanging around for some gnarly package incompatibility reasons that the yarn.lock reset has shaken out? Anyway, I'm trying to say, maybe the huge drop in bytesize is an anomaly.

One for WebEx to answer properly but my feeling is these files seem to come and go anyway, I've flagged something similar before and was told it was fine that we lose some and create others.

**Comment by joecowton1** [comment]
@jamesgorrie are you happy with us to merge this? Feels like we probably just need a voice from the the WebEx team to approve :)

**Review by mxdvl — APPROVED** [review]
🪄

**Comment by sndrs** [comment]
> either we continue with a package.lock thats a mess and doesn't allow us to add packages, or we do a hard reset and make a real effort to stay on top of things from here on.

i would say we should not be worried about doing a hard reset at any point – we just need to double check everything when we do. 

it's only designed to prevent _unexpected_ changes going out, e.g. a bug in a newer version of a transitive dependency . 

a lockfile absolutely is not the final word in compatibility - that's us. if we're happy with the change then we should make it.

**Comment by sndrs** [comment]
> if we're happy with the change then we should make it

ha! it's better to submit comments _before_ pr are merged 🤦

**Pull request merged**